From 4deb826429ac475f2029bdd9ca1eca2d3056114d Mon Sep 17 00:00:00 2001
From: zhangfuxin <fxzhang@ict.ac.cn>
Date: Mon, 5 Dec 2022 16:54:56 +0800
Subject: [PATCH] add UNWIND_DATA_REG define for loongarch64

---
 library/std/src/personality/gcc.rs | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/library/std/src/personality/gcc.rs b/library/std/src/personality/gcc.rs
index 5fc1b91a1c3..d00504d5bdb 100644
--- a/library/std/src/personality/gcc.rs
+++ b/library/std/src/personality/gcc.rs
@@ -76,6 +76,9 @@
 #[cfg(any(target_arch = "riscv64", target_arch = "riscv32"))]
 const UNWIND_DATA_REG: (i32, i32) = (10, 11); // x10, x11
 
+#[cfg(target_arch = "loongarch64")]
+const UNWIND_DATA_REG: (i32, i32) = (4, 5); // a0, a1
+
 // The following code is based on GCC's C and C++ personality routines.  For reference, see:
 // https://github.com/gcc-mirror/gcc/blob/master/libstdc++-v3/libsupc++/eh_personality.cc
 // https://github.com/gcc-mirror/gcc/blob/trunk/libgcc/unwind-c.c
-- 
2.34.1

