--- rustc-1.65.0-src-ori/src/bootstrap/lib.rs	2022-11-02 22:36:24.000000000 +0800
+++ rustc-1.65.0-src/src/bootstrap/lib.rs	2022-12-03 22:50:06.085773911 +0800
@@ -209,7 +209,7 @@
     (
         Some(Mode::Std),
         "target_arch",
-        Some(&["asmjs", "spirv", "nvptx", "nvptx64", "le32", "xtensa"]),
+        Some(&["asmjs", "spirv", "nvptx", "nvptx64", "le32", "xtensa", "loongarch64"]),
     ),
     /* Extra names used by dependencies */
     // FIXME: Used by rustfmt is their test but is invalid (neither cargo nor bootstrap ever set

