Upstream-Status: Pending
Signed-off-by: Fuxin Zhang <fxzhang@ict.ac.cn>
---
diff -urN openssl-3.0.5.ori/Configurations/10-main.conf openssl-3.0.5/Configurations/10-main.conf
--- openssl-3.0.5.ori/Configurations/10-main.conf	2022-11-01 01:47:36.629532756 +0000
+++ openssl-3.0.5/Configurations/10-main.conf	2022-11-01 01:41:39.423556741 +0000
@@ -818,6 +818,18 @@
         perlasm_scheme   => "linux64",
     },
 
+    # loongarch64 
+    "linux-loongarch64" => {
+        inherit_from     => [ "linux-generic64"],
+        perlasm_scheme   => "linux64",
+    },
+
+    # loongarch64 
+    "linux64-loongarch64" => {
+        inherit_from     => [ "linux-generic64"],
+        perlasm_scheme   => "linux64",
+    },
+
     #### IA-32 targets...
     #### These two targets are a bit aged and are to be used on older Linux
     #### machines where gcc doesn't understand -m32 and -m64
diff -urN openssl-3.0.5.ori/util/perl/OpenSSL/config.pm openssl-3.0.5/util/perl/OpenSSL/config.pm
--- openssl-3.0.5.ori/util/perl/OpenSSL/config.pm	2022-11-01 01:47:55.717852446 +0000
+++ openssl-3.0.5/util/perl/OpenSSL/config.pm	2022-11-01 01:45:39.907578548 +0000
@@ -681,6 +681,7 @@
       ],
       [ '.*86-.*-linux1',         { target => "linux-aout" } ],
       [ 'riscv64-.*-linux.',      { target => "linux64-riscv64" } ],
+      [ 'loongarch64-.*-linux.',  { target => "linux64-loongarch64" } ],
       [ '.*-.*-linux.',           { target => "linux-generic32" } ],
       [ 'sun4[uv].*-.*-solaris2',
         sub {
